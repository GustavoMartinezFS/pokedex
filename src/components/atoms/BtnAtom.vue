<template>
  <button
    :class="[
      'btn',
      disabled ? 'btn--disabled' : '',
      icon ? 'btn--with-icon' : 'btn--no-icon',
      `btn--${color}`
    ]"
    :style="{
      width: widthBt,
      maxWidth: `${maxWidthBt}px`,
      minWidth: `${minWidthBt}px`,
      height: `${height}px`,
      padding: `${padding}px`,
      margin: `${margin}px`,
      backgroundColor: isPressed ? pressedColor : active? color : disabledColor
    }"
    @click="handleClick"
    @mousedown="handleMouseDown"
    @mouseup="handleMouseUp"
    @mouseleave="handleMouseUp"
    :disabled="disabled"
  >
    <template v-if="icon">
      <span class="icon">
        <img :src="selectedIcon" alt="icon" />
      </span>
    </template>
    <span class="text">{{ name }}</span>
  </button>
</template>

<script lang="ts" setup>
import { defineProps, computed, ref } from 'vue';

const props = defineProps({
  color: {
    type: String,
    default: '#F22539',
  },
  disabledColor: {
    type: String,
    default: '#BFBFBF',
  },
  pressedColor: {
    type: String,
    default: '#C00E20',
  },
  active: {
    type: Boolean,
    default: false,
  },
  pressed: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  widthBt: {
    type: String,
    default: '100%',
  },
  minWidthBt: {
    type: Number,
    default: 131,
  },
  maxWidthBt: {
    type: Number,
    default: 275,
  },
  height: {
    type: Number,
    default: 44,
  },
  padding: {
    type: Number,
    default: 10,
  },
  margin: {
    type: String,
    default: '0',
  },
  name: {
    type: String,
    required: true,
  },
  icon: {
    type: String,
    default: null,
  },
});

const iconMap = {
  star: new URL('@/assets/star.svg', import.meta.url).href,
  list: new URL('@/assets/list.svg', import.meta.url).href,
};

const isPressed = ref(false);

const emit = defineEmits(['manage:click']);

const handleClick = () => {
  if (!props.disabled) {
    emit('manage:click');
  }
};

const handleMouseDown = () => {
  if (props.pressed) {
    isPressed.value = true;
  }
};

const handleMouseUp = () => {
  if (props.pressed) {
    isPressed.value = false;
  }
};

const selectedIcon = computed(() => iconMap[props.icon] || '');
</script>

<style scoped>
.btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 50px;
  cursor: pointer;
  transition: background-color 0.3s ease;
  box-sizing: border-box;
  border: none;
}

.btn--disabled {
  background-color: #d3d3d3;
  color: #a9a9a9;
  cursor: not-allowed;
}

.icon {
  margin-right: 10px;
}

.text {
  display: inline-block;
  font: 18px "Lato",sans-serif;
  color: var(--vt-c-white);
}
</style>
