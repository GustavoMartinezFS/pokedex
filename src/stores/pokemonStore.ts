import { defineStore } from 'pinia';
import { getPokemonList, getPokemonDetails } from '@/services/pokemonApi'

export const usePokemonStore = defineStore('pokemon', {
  state: () => ({
    pokemons: [] as Array<{ name: string; isFavorite: boolean }>,
    favorites: [] as Array<{ name: string; isFavorite: boolean }>,
    loading: false,
    currentView: 'all',
    limit: 100,
    offset: 0,
    allLoaded: false,
  }),

  actions: {
    async fetchPokemons() {
      if (this.allLoaded) return;
      this.loading = true;
      try {
        const response = await getPokemonList(this.limit, this.offset);
        if (response.results.length === 0) {
          this.allLoaded = true;
        } else {
          this.pokemons = [...this.pokemons, ...response.results.map((pokemon: { name: string; isFavorite: boolean; }) => ({
            ...pokemon,
            isFavorite: false,
          }))];
          this.offset += this.limit;
        }
      } catch (error) {
        console.error('Error fetching pokemons', error);
      } finally {
        this.loading = false;
      }
    },

    async fetchPokemonDetails(name: string) {
      this.loading = true;
      try{
        const response = await getPokemonDetails(name);
        const types = response.types.map((typeInfo: { type: { name: string; }; }) => (typeInfo.type.name));

        const pokemon = {
          name: response.name,
          height: response.height,
          weight: response.weight,
          types: types.join(', '),
          image: response.sprites.other.dream_world.front_default,
        }
        return pokemon;
      } catch (error) {
        console.error(`Error fetching details for ${name}:`, error);
      } finally {
        this.loading = false;
      }
    },

    addToFavorites(pokemon: {name: string; isFavorite: boolean }) {
      const indexFav = this.favorites.findIndex((fav) => fav.name === pokemon.name)

      const indexPok = this.pokemons.findIndex((fav) => fav.name === pokemon.name)

      if (!this.favorites.includes(pokemon)) {
        this.favorites.push(pokemon);
        this.pokemons[indexPok].isFavorite = true;
      } else {
        this.favorites.splice(indexFav,1);
        this.pokemons[indexPok].isFavorite = false;
      }
    },

    setView(view: string) {
      this.currentView = view;
    },
  }
});
