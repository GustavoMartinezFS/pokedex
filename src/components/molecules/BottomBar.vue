<template>
  <div class="bottom-bar">
    <btn-atom
      name="All"
      :maxWidthBt="275"
      :minWidthBt="150"
      :margin="buttonMargin"
      icon="list"
      :active="currentView === 'all'"
      @manage:click="setView('all')"
    />

    <div class="spacer"></div>

    <btn-atom
      name="Favorites"
      :maxWidthBt="275"
      :minWidthBt="150"
      :margin="buttonMargin"
      icon="star"
      :active="currentView === 'favorites'"
      @manage:click="setView('favorites')"
    />
  </div>
</template>

<script lang="ts" setup>
import { ref, onBeforeUnmount, onMounted } from 'vue';
import { usePokemonStore } from "@/stores/pokemonStore";
import { storeToRefs } from 'pinia';

import BtnAtom from "@/components/atoms/BtnAtom.vue";

const store = usePokemonStore();
const { currentView } = storeToRefs(store);
const buttonMargin = ref("0");

onMounted(() => {
  updateMargin();
  window.addEventListener("resize", updateMargin);
});

onBeforeUnmount(() => {
  window.removeEventListener("resize", updateMargin);
});

const setView = (view: string) => {
  store.setView(view);
};

const updateMargin = () => {
  buttonMargin.value = window.innerWidth < 600 ? "0 20px" : "0";
};
</script>

<style scoped>
.bottom-bar {
  display: flex;
  justify-content: center;
  padding: 10px;
  background-color: var(--c-dark-white);
  align-items: center;
  height: 80px;
  position: fixed;
  bottom: 0;
  width: 100%;
  border-top: 2px solid rgba(0, 0, 0, 0.1);
}

.spacer {
  width: 20px;
}
</style>
