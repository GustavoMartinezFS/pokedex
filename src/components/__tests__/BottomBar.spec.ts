import { mount } from '@vue/test-utils';
import { vi } from 'vitest';
import BottomBar from '@/components/molecules/BottomBar.vue';
import BtnAtom from '@/components/atoms/BtnAtom.vue';
import { setActivePinia, createPinia } from 'pinia';
import { usePokemonStore } from '@/stores/pokemonStore';

describe('BottomBar.vue', () => {
  let store: ReturnType<typeof usePokemonStore>;

  beforeEach(() => {
    setActivePinia(createPinia());
    store = usePokemonStore();
  });

  it('renders two BtnAtom components', () => {
    const wrapper = mount(BottomBar);
    const buttons = wrapper.findAllComponents(BtnAtom);

    expect(buttons.length).toBe(2);
    expect(buttons[0].props('name')).toBe('All');
    expect(buttons[1].props('name')).toBe('Favorites');
  });

  it('sets the correct view when buttons are clicked', async () => {
    const wrapper = mount(BottomBar);

    const allButton = wrapper.findComponent(BtnAtom);
    await allButton.vm.$emit('manage:click');

    expect(store.currentView).toBe('all');

    const favoritesButton = wrapper.findAllComponents(BtnAtom)[1];
    await favoritesButton.vm.$emit('manage:click');

    expect(store.currentView).toBe('favorites');
  });

  it('updates button margins on window resize', async () => {
    const wrapper = mount(BottomBar);
    const initialMargin = wrapper.vm.buttonMargin;

    window.innerWidth = 500;
    window.dispatchEvent(new Event('resize'));

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.buttonMargin).not.toBe(initialMargin);
    expect(wrapper.vm.buttonMargin).toBe('0 20px');
  });

  it('removes resize event listener on component unmount', () => {
    const removeEventListenerSpy = vi.spyOn(window, 'removeEventListener');
    const wrapper = mount(BottomBar);

    wrapper.unmount();
    expect(removeEventListenerSpy).toHaveBeenCalledWith('resize', expect.any(Function));
  });

  it('matches snapshot', () => {
    const wrapper = mount(BottomBar);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
