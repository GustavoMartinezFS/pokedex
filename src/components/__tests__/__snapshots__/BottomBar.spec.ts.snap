// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomBar.vue > matches snapshot 1`] = `
"<div data-v-448252c2="" class="bottom-bar"><button data-v-94a85ee2="" data-v-448252c2="" class="btn btn--with-icon btn--#F22539" style="width: 100%; max-width: 275px; min-width: 150px; height: 44px; padding: 10px; margin: 0px; background-color: rgb(242, 37, 57);"><span data-v-94a85ee2="" class="icon"><img data-v-94a85ee2="" src="http://localhost:3000/src/assets/list.svg" alt="icon"></span><span data-v-94a85ee2="" class="text">All</span></button>
  <div data-v-448252c2="" class="spacer"></div><button data-v-94a85ee2="" data-v-448252c2="" class="btn btn--with-icon btn--#F22539" style="width: 100%; max-width: 275px; min-width: 150px; height: 44px; padding: 10px; margin: 0px; background-color: rgb(191, 191, 191);"><span data-v-94a85ee2="" class="icon"><img data-v-94a85ee2="" src="http://localhost:3000/src/assets/star.svg" alt="icon"></span><span data-v-94a85ee2="" class="text">Favorites</span></button>
</div>"
`;
